javaparser-core/src/main/java/com/github/javaparser/ast/Node.java

```diff
@@ -677,4 +683,91 @@ protected SymbolResolver getSymbolResolver() {
     // We need to expose it because we will need to use it to inject the SymbolSolver
     public static final DataKey<SymbolResolver> SYMBOL_RESOLVER_KEY = new DataKey<SymbolResolver>() {
     };
+
+
+    public enum TreeTraversal {
+        PREORDER, BREADTHFIRST, POSTORDER
+    }
+
+    public Iterator<Node> treeIterator(TreeTraversal traversal) {
+        switch (traversal) {
+            case BREADTHFIRST:
+                return new TreeVisitor.BreadthFirstIterator(this);
+            case POSTORDER:
+                return new TreeVisitor.PostOrderIterator(this);
+            case PREORDER:
+                return new TreeVisitor.PreOrderIterator(this);
+            default:
+                throw new IllegalArgumentException("Unknown traversal choice.");
+        }
+    }
+    
+    public Iterable<Node> treeIterable(TreeTraversal traversal) {
+        return () -> treeIterator(traversal);
+    }
+    
+    public Stream<Node> treeStream(TreeTraversal traversal) {
+        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(treeIterator(traversal), NONNULL | DISTINCT), false);
+    }
+    
+    /**
+     * Walks the AST, calling the consumer for every node.
+     */
+    public void walk(Consumer<Node> consumer) {
+        treeIterable(TreeTraversal.PREORDER).forEach(consumer);
+    }
+
+    /**
+     * Walks the AST, calling the consumer for every node of type "nodeType".
+     */
+    public <T extends Node> void walk(Class<T> nodeType, Consumer<T> consumer) {
+        for (Node node : treeIterable(TreeTraversal.PREORDER)) {
+            if (nodeType.isInstance(node)) {
+                consumer.accept(nodeType.cast(node));
+            }
+        }
+    }
+
+    /**
+     * Walks the AST, returning all nodes of type "nodeType".
+     */
+    public <T extends Node> List<T> findAll(Class<T> nodeType) {
+        final List<T> found = new ArrayList<>();
+        walk(nodeType, found::add);
+        return found;
+    }
+
+    /**
+     * Walks the AST, returning all nodes of type "nodeType" that match the predicate.
+     */
+    public <T extends Node> List<T> findAll(Class<T> nodeType, Predicate<T> predicate) {
```

Comment:
Now we should rebrand JavaParser as the _functional_ parser for Java. It will bring us success and glory!

Reply:
I doubt we'll get away with having everything mutable ;-)